File: src/score/measures.ts
import type { GameNote, TimeSignatureEvent } from "../music/types";

export interface ScoreMeasureNote {
  note: GameNote;
  index: number;
  beat: number;
  durationBeats: number;
  offsetBeats: number;
}

export interface ScoreMeasure {
  number: number;
  startBeat: number;
  endBeat: number;
  lengthBeats: number;
  numerator: number;
  denominator: number;
  notes: ScoreMeasureNote[];
}

const EPSILON = 1e-6;

function noteBeat(note: GameNote, bpm: number): number {
  if (note.beat !== undefined) return note.beat;
  return note.start * bpm / 60000;
}

function noteDurationBeats(note: GameNote, bpm: number): number {
  if (note.durationBeats !== undefined) return note.durationBeats;
  return Math.max(1 / 96, note.duration * bpm / 60000);
}

function normalizeSignatures(events: TimeSignatureEvent[]): TimeSignatureEvent[] {
  const sorted = [...events]
    .filter((event) => event.numerator > 0 && event.denominator > 0)
    .sort((a, b) => a.beat - b.beat);
  if (sorted.length === 0 || sorted[0].beat > EPSILON) {
    sorted.unshift({ beat: 0, numerator: 4, denominator: 4 });
  }
  return sorted;
}

function signatureAt(beat: number, signatures: TimeSignatureEvent[]): TimeSignatureEvent {
  let current = signatures[0];
  for (let index = 1; index < signatures.length; index += 1) {
    if (signatures[index].beat > beat + EPSILON) break;
    current = signatures[index];
  }
  return current;
}

function nextSignatureAfter(beat: number, signatures: TimeSignatureEvent[]): TimeSignatureEvent | undefined {
  return signatures.find((event) => event.beat > beat + EPSILON);
}

function normalizedStarts(starts: number[], maxBeat: number): number[] {
  const result = [...starts]
    .filter((beat) => Number.isFinite(beat) && beat >= 0 && beat <= maxBeat + EPSILON)
    .sort((a, b) => a - b)
    .filter((beat, index, array) => index === 0 || Math.abs(beat - array[index - 1]) > EPSILON);
  if (result.length > 0 && result[0] > EPSILON) result.unshift(0);
  return result;
}

function deriveStarts(maxBeat: number, signatures: TimeSignatureEvent[]): number[] {
  const starts = [0];
  let cursor = 0;
  let guard = 0;
  while (cursor < maxBeat - EPSILON && guard < 10000) {
    guard += 1;
    const signature = signatureAt(cursor, signatures);
    const nominalLength = signature.numerator * 4 / signature.denominator;
    const nextSignature = nextSignatureAfter(cursor, signatures);
    let next = cursor + nominalLength;
    if (nextSignature && nextSignature.beat < next - EPSILON) next = nextSignature.beat;
    if (next <= cursor + EPSILON) next = cursor + Math.max(1 / 16, nominalLength);
    cursor = next;
    if (cursor < maxBeat - EPSILON) starts.push(cursor);
  }
  return starts;
}

export function buildScoreMeasures(
  notes: GameNote[],
  timeSignatures: TimeSignatureEvent[],
  bpm: number,
  explicitMeasureStarts: number[] = []
): ScoreMeasure[] {
  if (notes.length === 0) return [];
  const safeBpm = bpm > 0 ? bpm : 120;
  const signatures = normalizeSignatures(timeSignatures);
  const noteData = notes.map((note, index) => {
    const beat = noteBeat(note, safeBpm);
    const durationBeats = noteDurationBeats(note, safeBpm);
    return { note, index, beat, durationBeats };
  });
  const maxBeat = Math.max(...noteData.map((item) => item.beat + item.durationBeats));

  let starts = normalizedStarts(explicitMeasureStarts, maxBeat);
  if (starts.length === 0) starts = deriveStarts(maxBeat, signatures);
  else {
    let cursor = starts[starts.length - 1];
    let guard = 0;
    while (cursor < maxBeat - EPSILON && guard < 10000) {
      guard += 1;
      const signature = signatureAt(cursor, signatures);
      const nominalLength = signature.numerator * 4 / signature.denominator;
      const next = cursor + nominalLength;
      if (next >= maxBeat - EPSILON) break;
      cursor = next;
      starts.push(cursor);
    }
  }

  return starts.map((startBeat, measureIndex) => {
    const signature = signatureAt(startBeat, signatures);
    const nominalLength = signature.numerator * 4 / signature.denominator;
    const nextExplicit = starts[measureIndex + 1];
    const endBeat = nextExplicit ?? Math.max(startBeat + nominalLength, maxBeat);
    const lengthBeats = Math.max(1 / 16, endBeat - startBeat);
    const measureNotes = noteData
      .filter((item) => item.beat >= startBeat - EPSILON && item.beat < endBeat - EPSILON)
      .map((item) => ({ ...item, offsetBeats: Math.max(0, item.beat - startBeat) }));

    return {
      number: measureIndex + 1,
      startBeat,
      endBeat,
      lengthBeats,
      numerator: signature.numerator,
      denominator: signature.denominator,
      notes: measureNotes
    };
  });
}

export function durationLabel(beats: number): string {
  const value = Math.max(0, beats);
  const candidates: Array<[number, string]> = [
    [4, "1"], [3, "2·"], [2, "2"], [1.5, "4·"], [1, "4"],
    [0.75, "8·"], [0.5, "8"], [0.375, "16·"], [0.25, "16"], [0.125, "32"]
  ];
  const match = candidates.find(([candidate]) => Math.abs(candidate - value) < 0.04);
  return match ? match[1] : `${Math.round(value * 100) / 100}b`;
}

File: desktop/DFH.Core/Score/Measures.cs
using DFH.Core.Harmonica;
using DFH.Core.Music;

namespace DFH.Core.Score;

public sealed record ScoreMeasureNote(GameNote Note, int Index, double Beat, double DurationBeats, double OffsetBeats);

public sealed record ScoreMeasure(
    int Number,
    double StartBeat,
    double EndBeat,
    double LengthBeats,
    int Numerator,
    int Denominator,
    IReadOnlyList<ScoreMeasureNote> Notes);

/// <summary>Port of buildScoreMeasures in src/score/measures.ts, used by the score viewer.</summary>
public static class Measures
{
    private const double Epsilon = 1e-6;

    private static double NoteBeat(GameNote note, double bpm) => note.Beat ?? note.Start * bpm / 60000;

    private static double NoteDurationBeats(GameNote note, double bpm) =>
        note.DurationBeats ?? Math.Max(1.0 / 96, note.Duration * bpm / 60000);

    private static List<TimeSignatureEvent> NormalizeSignatures(IReadOnlyList<TimeSignatureEvent> events)
    {
        var sorted = events.Where(e => e.Numerator > 0 && e.Denominator > 0).OrderBy(e => e.Beat).ToList();
        if (sorted.Count == 0 || sorted[0].Beat > Epsilon) sorted.Insert(0, new TimeSignatureEvent(0, 4, 4));
        return sorted;
    }

    private static TimeSignatureEvent SignatureAt(double beat, List<TimeSignatureEvent> signatures)
    {
        var current = signatures[0];
        for (var index = 1; index < signatures.Count; index += 1)
        {
            if (signatures[index].Beat > beat + Epsilon) break;
            current = signatures[index];
        }
        return current;
    }

    private static TimeSignatureEvent? NextSignatureAfter(double beat, List<TimeSignatureEvent> signatures) =>
        signatures.FirstOrDefault(e => e.Beat > beat + Epsilon);

    private static List<double> NormalizedStarts(IReadOnlyList<double> starts, double maxBeat)
    {
        var result = starts
            .Where(beat => double.IsFinite(beat) && beat >= 0 && beat <= maxBeat + Epsilon)
            .OrderBy(beat => beat)
            .ToList();
        var deduped = new List<double>();
        for (var index = 0; index < result.Count; index += 1)
        {
            if (index == 0 || Math.Abs(result[index] - result[index - 1]) > Epsilon) deduped.Add(result[index]);
        }
        if (deduped.Count > 0 && deduped[0] > Epsilon) deduped.Insert(0, 0);
        return deduped;
    }

    private static List<double> DeriveStarts(double maxBeat, List<TimeSignatureEvent> signatures)
    {
        var starts = new List<double> { 0 };
        double cursor = 0;
        var guard = 0;
        while (cursor < maxBeat - Epsilon && guard < 10000)
        {
            guard += 1;
            var signature = SignatureAt(cursor, signatures);
            var nominalLength = signature.Numerator * 4.0 / signature.Denominator;
            var nextSignature = NextSignatureAfter(cursor, signatures);
            var next = cursor + nominalLength;
            if (nextSignature != null && nextSignature.Beat < next - Epsilon) next = nextSignature.Beat;
            if (next <= cursor + Epsilon) next = cursor + Math.Max(1.0 / 16, nominalLength);
            cursor = next;
            if (cursor < maxBeat - Epsilon) starts.Add(cursor);
        }
        return starts;
    }

    public static IReadOnlyList<ScoreMeasure> Build(
        IReadOnlyList<GameNote> notes,
        IReadOnlyList<TimeSignatureEvent> timeSignatures,
        double bpm,
        IReadOnlyList<double>? explicitMeasureStarts = null)
    {
        if (notes.Count == 0) return [];
        var safeBpm = bpm > 0 ? bpm : 120;
        var signatures = NormalizeSignatures(timeSignatures);
        var noteData = notes
            .Select((note, index) => (Note: note, Index: index, Beat: NoteBeat(note, safeBpm), DurationBeats: NoteDurationBeats(note, safeBpm)))
            .ToList();
        var maxBeat = noteData.Max(item => item.Beat + item.DurationBeats);

        var starts = NormalizedStarts(explicitMeasureStarts ?? [], maxBeat);
        if (starts.Count == 0)
        {
            starts = DeriveStarts(maxBeat, signatures);
        }
        else
        {
            var cursor = starts[^1];
            var guard = 0;
            while (cursor < maxBeat - Epsilon && guard < 10000)
            {
                guard += 1;
                var signature = SignatureAt(cursor, signatures);
                var nominalLength = signature.Numerator * 4.0 / signature.Denominator;
                var next = cursor + nominalLength;
                if (next >= maxBeat - Epsilon) break;
                cursor = next;
                starts.Add(cursor);
            }
        }

        var measures = new List<ScoreMeasure>(starts.Count);
        for (var measureIndex = 0; measureIndex < starts.Count; measureIndex += 1)
        {
            var startBeat = starts[measureIndex];
            var signature = SignatureAt(startBeat, signatures);
            var nominalLength = signature.Numerator * 4.0 / signature.Denominator;
            var endBeat = measureIndex + 1 < starts.Count ? starts[measureIndex + 1] : Math.Max(startBeat + nominalLength, maxBeat);
            var lengthBeats = Math.Max(1.0 / 16, endBeat - startBeat);
            var measureNotes = noteData
                .Where(item => item.Beat >= startBeat - Epsilon && item.Beat < endBeat - Epsilon)
                .Select(item => new ScoreMeasureNote(item.Note, item.Index, item.Beat, item.DurationBeats, Math.Max(0, item.Beat - startBeat)))
                .ToList();
            measures.Add(new ScoreMeasure(measureIndex + 1, startBeat, endBeat, lengthBeats, signature.Numerator, signature.Denominator, measureNotes));
        }
        return measures;
    }

    public static string DurationLabel(double beats)
    {
        var value = Math.Max(0, beats);
        (double Beats, string Label)[] candidates =
        [
            (4, "1"), (3, "2·"), (2, "2"), (1.5, "4·"), (1, "4"),
            (0.75, "8·"), (0.5, "8"), (0.375, "16·"), (0.25, "16"), (0.125, "32")
        ];
        foreach (var (candidate, label) in candidates)
        {
            if (Math.Abs(candidate - value) < 0.04) return label;
        }
        return $"{Math.Round(value * 100) / 100}b";
    }
}
